### ISSUE TITLE
No initial write access to GPX export folder

### ISSUE THREAD

**Issue by alan666notb** [cgeo/cgeo#10782]
## Bug description 

I bought a new phone a couple of weeks ago, I installed cgeo and it went through the "SAF Wizard" to give access to the necessary folders. Everything worked including importing GPX files that I generated from GSAK.

Yesterday, I tried to export a GPX file to transfer to another phone and it failed (Popup message was just something like "Export failed").

I eventually solved it by going to settings/offline data/GPX and re-selecting my GPX folder. It then asked me for access to that folder and after that the export worked.

It looks like the initial setup/wizard doesn't give write access to that folder (But does allow read access).

## Reproducible 
I can't test this without uninstalling/re-installing, which I don't want to do.

## System information 

```
--- System information ---
c:geo version: 2021.05.20

Device:
-------
- Device type: M2101K6G (sweet_eea, Redmi)
- Available processors: 8
- Android version: 11
- Android build: RKQ1.200826.002 test-keys
- Sailfish OS detected: false
- Google Play services: enabled - 21.18.16 (150400-374723149)
- HW acceleration: enabled (default state)

Sensor and location:
-------
- Low power mode: inactive
- Compass capabilities: yes
- Rotation vector sensor: present
- Orientation sensor: present
- Magnetometer & Accelerometer sensor: present
- Direction sensor used: rotation vector

Program settings:
-------
- Hide caches: archived
- Hide waypoints: -
- Set language: en_GB (system default)
- System date format: dd/MM/y
- Debug mode active: no
- Live map mode: false
- OSM multi-threading: true / threads: 4
- Global filter: display all caches
- Last backup: 26 May, 15:26
- Routing mode: Off
 -Settings: v5, Count:191
- Map: Great_Britain_OAM_V4_May2021 (Offline)
  - Id: cgeo.geocaching.maps.mapsforge.MapsforgeMapProvider$OfflineMapSource:primary:OpenStreetMaps/Great_Britain_OAM_V4_May2021.map
  - Atts: Maps: (c)Christian Kernbeis, Themes: (c)Tobias Kuehn, www.openandromaps.org 2012-2021
  - Theme: ElevateTheme.zip:Elevate.xml

Services:
-------
- Geocaching sites enabled:
   geocaching.com: Logged in (Login OK) / PREMIUM
- Geocaching.com date format: dd MMM yy
- Routing: external / BRouter installed: false
- Installed c:geo plugins:  none

Permissions & paths:
-------
- Fine location permission: granted
- Write external storage permission: granted
- System internal c:geo dir: /data/user/0/cgeo.geocaching (59.9 GB free) v2 internal isDir(6 entries)
- Legacy User storage c:geo dir: /storage/emulated/0/cgeo (59.9 GB free) v2 external non-removable isDir(5 entries)
- Geocache data: /storage/emulated/0/Android/data/cgeo.geocaching/files/GeocacheData (59.9 GB free) v2 external non-removable isDir(781 entries)
- Internal theme sync (is turned off): /data/user/0/cgeo.geocaching/MapThemeData (59.9 GB free) v2 internal isDir(0 entries)
- Public Folders: #11
  - BASE: /cgeo (User-Defined)[/cgeo[DOCUMENT#0:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo, Av:true, files:>=6, dirs:>=10, totalFileSize:>=99.1 MB, free space: 59.9 GB, files on device: 13511677)
  - OFFLINE_MAPS: /OpenStreetMaps (User-Defined)[/OpenStreetMaps[DOCUMENT#0:p-content://com.android.externalstorage.documents/tree/primary%3AOpenStreetMaps::]] (Uri: content://com.android.externalstorage.documents/tree/primary%3AOpenStreetMaps/document/primary%3AOpenStreetMaps, Av:true, files:4, dirs:0, totalFileSize:3.2 GB, free space: 59.9 GB, files on device: 13511677)
  - OFFLINE_MAP_THEMES: /OpenStreetMaps (User-Defined)[/OpenStreetMaps[DOCUMENT#0:p-content://com.android.externalstorage.documents/tree/primary%3AOpenStreetMaps::]] (Uri: content://com.android.externalstorage.documents/tree/primary%3AOpenStreetMaps/document/primary%3AOpenStreetMaps, Av:true, files:4, dirs:0, totalFileSize:3.2 GB, free space: 59.9 GB, files on device: 13511677)
  - LOGFILES: /cgeo/logfiles (Default)[/cgeo/logfiles[PERSISTABLE_FOLDER(BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/logfiles]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2Flogfiles, Av:true, files:0, dirs:0, totalFileSize:0 B, free space: 59.9 GB, files on device: 13511677)
  - GPX: /My Documents/SyncedWithGoogle/GPX (User-Defined)[/My Documents/SyncedWithGoogle/GPX[DOCUMENT#0:p-content://com.android.externalstorage.documents/tree/primary%3AMy%20Documents%2FSyncedWithGoogle%2FGPX::]] (Uri: content://com.android.externalstorage.documents/tree/primary%3AMy%20Documents%2FSyncedWithGoogle%2FGPX/document/primary%3AMy%20Documents%2FSyncedWithGoogle%2FGPX, Av:true, files:22, dirs:2, totalFileSize:23.5 MB, free space: 59.9 GB, files on device: 13511677)
  - BACKUP: /cgeo/backup (Default)[/cgeo/backup[PERSISTABLE_FOLDER(BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/backup]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2Fbackup, Av:true, files:>=10, dirs:>=5, totalFileSize:>=166.6 MB, free space: 59.9 GB, files on device: 13511677)
  - FIELD_NOTES: /cgeo/field-notes (Default)[/cgeo/field-notes[PERSISTABLE_FOLDER(BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/field-notes]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2Ffield-notes, Av:true, files:0, dirs:0, totalFileSize:0 B, free space: 59.9 GB, files on device: 13511677)
  - SPOILER_IMAGES: /cgeo/GeocachePhotos (Default)[/cgeo/GeocachePhotos[PERSISTABLE_FOLDER(BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/GeocachePhotos]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2FGeocachePhotos, Av:true, files:0, dirs:0, totalFileSize:0 B, free space: 59.9 GB, files on device: 13511677)
  - ROUTING_BASE: /cgeo/routing (Default)[/cgeo/routing[PERSISTABLE_FOLDER(BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/routing]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2Frouting, Av:true, files:7, dirs:1, totalFileSize:81.2 KB, free space: 59.9 GB, files on device: 13511677)
  - ROUTING_TILES: /cgeo/routing/segments4 (Default)[/cgeo/routing/segments4[PERSISTABLE_FOLDER(ROUTING_BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/routing/segments4]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2Frouting%2Fsegments4, Av:true, files:0, dirs:0, totalFileSize:0 B, free space: 59.9 GB, files on device: 13511677)
  - TEST_FOLDER: [Legacy]/data/user/0/cgeo.geocaching/files/unittest (Default)[/data/user/0/cgeo.geocaching/files/unittest[FILE#1:p-file:///data/user/0/cgeo.geocaching/files::/unittest]] (Uri: file:///data/user/0/cgeo.geocaching/files/unittest, Av:true, files:0, dirs:0, totalFileSize:0 B, free space: 59.9 GB, files on device: -1)
- Map render theme path: ElevateTheme.zip:Elevate.xml
- PersistedDocumentUris: #1
- TRACK: null
- Persisted Uri Permissions: #3
  - content://com.android.externalstorage.documents/tree/primary%3AMy%20Documents%2FSyncedWithGoogle%2FGPX (26 May, 15:31):RW
  - content://com.android.externalstorage.documents/tree/primary%3AOpenStreetMaps (7 May, 19:56):RW
  - content://com.android.externalstorage.documents/tree/primary%3Acgeo (7 May, 19:45):RW
- Database: /storage/emulated/0/Android/data/cgeo.geocaching/files/databases/data (v94, Size:34.7 MB) on user storage
--- End of system information ---
```

**Comment by eddiemuc** [comment]
Hi @alan666notb your Settings look ok now. Of course I don't know how they looked before (when you had the problem)

The wizard and the Settings use the same routines to get folder access rights, and c:geo always asks for write access (never for just read-only). I am currently at a loss what caused the experienced behavior. Is the GPX folder on an SD card or internal memory? Anything else which might make this folder "special" when compared to e.g. the base `cgeo` folder?

If this happens again, it would be great to get Settings (while problem exists) as well as a log file.

**Comment by alan666notb** [comment]
I can't think of anything unusual about the GPX folder - it's on internal storage, same as the maps & base folder.
Are there any circumstances where cgeo writes to the maps or map themes folders?

**Comment by eddiemuc** [comment]
> Are there any circumstances where cgeo writes to the maps or map themes folders?

Yes, it is doing this when you download offline maps/themes via c:geo. Might be an interesting thing to try.

**Comment by alan666notb** [comment]
Tried downloading a map: it worked fine and copied it to my 'maps' folder.

Does the wizard actually ask for access to the GPX folder - now that I think about it, I don't recall it doing so.
I remember it asking for access to the base (/cgeo), maps and map themes folders, but not GPX.

**Comment by eddiemuc** [comment]
I think the wizard does not ask explicitly for GPX folder but rather sets the default for it (which is a subfolder "gpx" under the selected base folder)

**Comment by moving-bits** [comment]
The wizard itself does not ask you for a GPX folder, as it wants to ask as few questions as possible, and cannot know on a fresh install whether you ever want to use the GPX import/export. It should be able to use the default folder in this case (if you have granted access to the base folder).

It's a bit different if you have set a GPX folder in your old installation and you import their settings - then c:geo should ask you for re-granting access to the configured folder.

**Comment by alan666notb** [comment]
> It's a bit different if you have set a GPX folder in your old installation and you import their settings - then c:geo should ask you for re-granting access to the configured folder.

Yes - that's what I did. I copied a backup file from my old phone to the new one and imported it.

My GPX folder isn't under the base folder, so I guess it just pointed it to the correct folder (Same name on both phones) and left it like that without asking permission.

But - I've been able to import GPX files from that folder, shouldn't Android-11 prevent that without specific permission?

**Comment by moving-bits** [comment]
But on importing a settings file c:geo should ask you for access to the folder. Strange...

**Comment by eddiemuc** [comment]
Ok with this additional information I have a. Idea what has happened: until recently there was a bug in cgeo which prevented the re-granting of folder access to be executed when a restore was done via installation wizard. This issue is already fixed, but maybe it was still in the c:geo version where @alan666notb initially imported the old settings?

If this is the cause of the problem then it is already fixed.

**Comment by alan666notb** [comment]
I would have setup my new phone on 7th May,
I'm on the beta program, would the fix have been released since then?

**Comment by eddiemuc** [comment]
PR was #10706, it fixed issue #10690.

It was merged on 22.5., so I guess it was not included in your version. 

I am somewhat confident that this was the source of your problem.

**Comment by alan666notb** [comment]
OK, I'll close this.
Thanks.
